Allow InfluxDB raw queries

drlibre opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add the capability to create complex queries sensors with InfluxDB API 1.XX or API 2.xx by using a 'raw' format of query in the configuration parameters of the sensor.
With the standard 'queries' configuration of the component, it's only possible, for instance, to use 1 group function in API 1.xx mode. If one need to use two functions (for instance to get the sum of the max of measurement values), it is at least tricky and sometime not possible. 
This add bye the way the capability to define the query for the component just with a copy & paste of the query used in the InfluxDB UI which can be easier for some users.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
One of my use case:
I need a sensor that get from InfluxDB (API version 1.xx) the summ (from a date defined with an input_datetime sensor) of the maximum value of the day of a sensor. 
The query is something like:

```sql
SELECT sum("max") FROM (SELECT max("value") FROM "HomeAssistant"."autogen"."l" 
WHERE time >= '{{ states('input_datetime.date_fuel_plein') }}T12:00:00+01:00' AND "entity_id"='bruleur_conso_j'
GROUP BY time(1d) )
```

Such a query is not possible for now with InfluxDB component. It would be (if the 'measurement' field type was 'Template' and not 'String') by defining in the sensor configuration:

```yaml
  group_function: sum
  field: '"max"'
  measurement: >-
      (SELECT max("value") FROM "HomeAssistant"."autogen"."l" 
      WHERE time >= '{{ states('input_datetime.date_fuel_plein') }}T12:00:00+01:00' AND 
      "entity_id"=''bruleur_conso_j'' GROUP BY time(1d))
  where: 'time>=0'
```

which isn't very user friendly and suppose that the type of the 'measurement' parameter is modified from 'string' to 'template'.

An other way to solve the problem would be to add an option to calculate an agregate value when the query returns multiple results, but why doing in the component code what InfluxDB is doing very well?...  

The proposed change consist in defining a new 'queries_raw' configuration type which seems to be the best option.

This 'queries_raw' uses the parameters:

```yaml
name: string, REQUIRED, The name of the sensor.
unit_of_measurement: string, (optional), Defines the units of measurement of the sensor, if any.
value_template: template, (optional), Defines a template to extract a value from the payload.
database: string, (optional, default: home_assistant), Name of the database to use in V1 mode.
bucket: string, (optional, default: Home Assistant), Name of the bucket within your Organization to read from.

query: template, REQUIRED, The query to use.
field: string, (optional), default: _value, The field name of the result of the query.
```

For the above use case, the configuration of the component would be:
```yaml
queries_raw:
  - name: "bruleur_conso_plein_test"
    database: HomeAssistant
    unit_of_measurement: "l"
    value_template: '{{ value | round(1) }}'
    query: >-
      SELECT sum("max") as sum FROM (SELECT max("value") FROM "HomeAssistant"."autogen"."l" 
      WHERE time >= '{{ states('input_datetime.date_fuel_plein') }}T12:00:00+01:00' AND "entity_id"='bruleur_conso_j' 
      GROUP BY time(1d) )
    field: sum
```
It is possible to mix "queries" and "queries_raw" in the component configuration for the defined API version. 

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/21272

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

drlibre committed:
Update const.py

Add 'CONF_QUERIES_RAW' and 'LANGUAGE_INFLUXQL_RAW' in order to manage the new 'queries_raw' option of the component

drlibre committed:
Add 'queries_raw' feature

Add a 'queries_raw' option for the component in order to use a "raw" query (just have to copy and paste from InfluxDB as a template)

homeassistant commented:
Hi @drlibre, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#63147).

Once you do that we will be able to review and accept this pull request.

Thanks!

drlibre committed:
Update sensor.py

mdegat01 commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -139,7 +153,8 @@ def validate_query_format_for_version(conf: dict) -> dict:
         {
             vol.Exclusive(CONF_QUERIES, "queries"): [_QUERY_SCHEMA[LANGUAGE_INFLUXQL]],
-            vol.Exclusive(CONF_QUERIES_FLUX, "queries"): [_QUERY_SCHEMA[LANGUAGE_FLUX]],
+            vol.Exclusive(CONF_QUERIES_RAW, "queries"): [_QUERY_SCHEMA[LANGUAGE_INFLUXQL_RAW]],
```
I don't think `queries_raw` should be exclusive here. By defining it this way you make it so users must specify all their queries in raw or none of their queries in raw. Why wouldn't a user be able to mix and match? Specify one complex query as raw but the others as normal using `queries` or `queries_flux`? I feel like the schema should only require that `CONF_QUERIES` and `CONF_QUERIES_FLUX` are exclusive since an Influx deployment can only accept queries in one language. `CONF_QUERIES_RAW` should be be separate and not exclusive with either.

The schema should also validate that at least one of `CONF_QUERIES`, `CONF_QUERIES_FLUX` and `CONF_QUERIES_RAW` is provided.

drlibre replied:
I agree that it's better to be able to mix... I'm looking for improving the proposed changes on that point.

Is there a proper way to do the the validation of configuration if set to optional (ie QUERIES and QUERIES_FLUX cannot be set together)?

mdegat01 replied:
Yea just don't bundle the new one into the `exclusive`. `vol.Exclusive` means that one one of the things in the named exclusive (`queries` in this case) can exist. Since `queries_raw` can co-exist with the others you just make it a normal `vol.Optional`. So it would be like this:

```python
    SENSOR_PLATFORM_SCHEMA.extend(COMPONENT_CONFIG_SCHEMA_CONNECTION).extend(
        {
            vol.Exclusive(CONF_QUERIES, "queries"): [_QUERY_SCHEMA[LANGUAGE_INFLUXQL]],
            vol.Optional(CONF_QUERIES_RAW): [_QUERY_SCHEMA[LANGUAGE_INFLUXQL_RAW]],
            vol.Exclusive(CONF_QUERIES_FLUX, "queries"): [_QUERY_SCHEMA[LANGUAGE_FLUX]],            
        }
    ),
```

drlibre replied:
Taken into account in the e7dfb9f commit

mdegat01 commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -48,6 +49,7 @@
     DEFAULT_RANGE_STOP,
     INFLUX_CONF_VALUE,
     INFLUX_CONF_VALUE_V2,
+    LANGUAGE_INFLUXQL_RAW,
```
Why is this `INFLUXQL_RAW`? Shouldn't you also be able to specify a flux query in the raw as well? It's raw so its just text either way.

drlibre replied:
Changed to `LANGUAGE_RAW` in the e7dfb9f commit. Used to define the "raw" type of configuration that can be used in V1 or V2 mode

mdegat01 commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -204,6 +226,15 @@ def __init__(self, hass, influx, query):
                 query.get(CONF_GROUP_FUNCTION),
             )
 
+        if query[CONF_LANGUAGE] == LANGUAGE_INFLUXQL_RAW:
```
```suggestion
        elif query[CONF_LANGUAGE] == LANGUAGE_INFLUXQL_RAW:
```
This code will break all InfluxV2 users right now since you've made this a separate if block with an else. For InfluxV2 users it will fall to the else and try and make an `InfluxQLSensorData` for each item in `queries_flux` which won't work.

drlibre replied:
Right, basic mistake, thanks ...

mdegat01 commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -123,6 +130,13 @@ def validate_query_format_for_version(conf: dict) -> dict:
+        {
+            vol.Optional(CONF_DB_NAME): cv.string,
+            vol.Required(CONF_QUERY): cv.template,
+            vol.Required(CONF_FIELD): cv.string,
```
```suggestion
            vol.Optional(CONF_FIELD, default=DEFAULT_FIELD): cv.string,
```
Think we should use the same default value here. That keeps it consistent. Also making this field optional allows this to be used for V2 Flux queries since specifying a field is something that only makes sense with InfluxQL queries. When talking to V2 systems `field` and `db_name` would just be ignored.

mdegat01 replied:
TBH you could probably just omit this from the config entirely. Just instruct users to put `as value` in their select statement next to whatever they are selecting to give it the expected name.

drlibre replied:
I thought it was more flexible and robust to set the name of the result of the query as a paraeter of configuration wether using the default 'value' which has in this case to be taken into account in the query.
And it may be simpler to extend to V2 Flux queries as I have in mind that default value for V2 is '_value'?

mdegat01 replied:
In V2 flux queries it is always `_value`. There's not an option. It's not SQL-like like it is with InfluxQL. Each each attribute of each entity ends up in its own table where every row has a `_time` column specifying the time that row was created and a `_value` column specifying the value at that time. The only other columns are what you designate as keys (tags in the config).

To show you what I mean, here's an example entity in my Home Assistant:
![Screen Shot 2022-01-05 at 2 12 11 PM](https://user-images.githubusercontent.com/2037026/148275081-9fa5e2ba-433d-4556-8dda-be95cc992b8f.png)

Here's an example flux query on my data about this entity as Influx sees it:
```
from(bucket: "michael.degatano's Bucket")
  |> range(start: -1h, stop: now())
  |> filter(fn: (r) => r.domain == "sensor" and r.entity_id == "family_room_couch_outlet")
```
And here's the result:
![Screen Shot 2022-01-05 at 2 10 13 PM](https://user-images.githubusercontent.com/2037026/148274916-0d50e5ed-ca0d-4d69-a0d0-a638a5fb4ad4.png)

Notice how in the left there is a separate table for each attribute you see in that entity. And then within each row of those tables there's really only one column that matters - `_value`. Everything else is just things that make up the key of that entity and the same in every row or a timestamp dating the row.

That's why it has to be optional for Influx2, there is no reason to specify a field here. The field you want will always be `_value`.

drlibre replied:
Thanks for those precisions and explanations about Influx2: no need so to get a field name for the result, just have to keep the `INFLUX_CONF_VALUE_V2`...

drlibre replied:
I saw that with InfluxV2 columns can possibly be renamed with the rename function (`rename(columns: {_value: "result"})` for instance).
Probably very unsual in the context, but the ability to specify a name for the result of the query seems to be the most flexible even in Influx2 mode.
The best solution seems to be to keep this parameter optional in both mode with default value as `_value` (which match the Influx2 default). For InfluxQL, user must specify either the name of the result in the query (and so just have to specify the default configuration value `_value` in the query for the result) or specify the name used in the query (and so specify a `field` in the configuration).
That's how it is implemented in the e7dfb9f commit.

mdegat01 reviewed (changes requested):
I think this idea makes sense. The raw option mimics what is provided for the [SQL Integration](https://www.home-assistant.io/integrations/sql/) which is good and consistent.

Although I don't like that its restricted to only work on V1 Influx. That's confusing since Flux queries are also just text. As a user I would expect a raw option to just allow me to copy and paste a raw query from my Influx UI and work, regardless of whether my Influx deployment was V1 or V2. I think `InfluxRawSensorData` should accept the language as an input and use that to handle the response accordingly. It's mostly identical since the `InfluxClient` returned by `get_influx_connection` already normalizes the read APIs of the two libraries. Just need to either grab `value` from the first row (V2) or do what you're doing right now (V1).

Also I don't see any tests in here. I can't approve the workflows to run, we'll have to wait for someone else for that. But I'm pretty sure once they do its just going to fail the code coverage test since all this new code won't be covered.

If you were waiting to see if the idea was approved before working more that's fair but we'll need tests of the new functionality to merge this.

mdegat01 commented:
Had a thought. `LANGUAGE_INFLUXQL_RAW` doesn't actually make sense because there isn't a third Influx language here. The language is either `Flux` or `influxQL`. Raw is simply another way to specify a query in one of those languages.

So instead of having a third language and a third type of sensor (`InfluxRawSensorData`) you could simply use `CONF_API_VERSION` to set `CONF_LANGUAGE` to either `LANGUAGE_INFLUXQL` or `LANGUAGE_FLUX` for each item in `queries_raw` in `validate_query_format_for_version`. Then modify `InfluxQLSensorData` and `InfluxFluxSensorData` to accept `raw_query` as a parameter and change how they set the query according to which parameters are specified.

1. If `raw_query` is provided just render it as a template and run it.
2. If `raw_query` is omitted then construct the query from other inputs like its doing today.

I think that might be easier. It's really just a difference in how the query is constructed, processing its results into a sensor is unchanged.

drlibre commented:
Many thanks for your comments.

> Although I don't like that its restricted to only work on V1 Influx.

I had in mind at the begining to achieve a solution without this restriction but didn't find the right way to achieve this.
Your comments helped me to explore more deeply some options.

> `LANGUAGE_INFLUXQL_RAW` doesn't actually make sense because there isn't a third Influx language here. The language is either `Flux` or `influxQL`. Raw is simply another way to specify a query in one of those languages.
>
> So instead of having a third language and a third type of sensor (`InfluxRawSensorData`) you could simply use `CONF_API_VERSION` to set `CONF_LANGUAGE` to either `LANGUAGE_INFLUXQL` or `LANGUAGE_FLUX` for each item in queries_raw in validate_query_format_for_version. Then modify `InfluxQLSensorData` and `InfluxFluxSensorData` to accept raw_query as a parameter and change how they set the query according to which parameters are specified.

I agree that `LANGUAGE_INFLUXQL_RAW` isn't a third influx language... It's more a type of configuration in that case. 
I considered how to improve the code taking into account your suggestions.
I think it would be simpler and clearer to keep  a `InfluxRawSensorData` type that would be able to process V1 or V2 Influx depending of `api_version` parameter.
I look to improve the proposed feature with some additional changes soon...

> Also I don't see any tests in here.

Sorry, I don't know exactly how to provide this...

mdegat01 commented:
> Sorry, I don't know exactly how to provide this...

All the tests for the InfluxDB integration can be found [here](https://github.com/home-assistant/core/tree/dev/tests/components/influxdb). Take a look at what's in there, you can use that as a basis to start from. Essentially you just want to feed configuration scenarios and then test that the sensors were set up the way you want and the result is what you expect.

If you aren't familiar with testing in python using `pytest` I'd recommend dropping into the [HA discord](https://discord.gg/c5DvZ4e). There's a `devs_core` channel where you can get help from other developers on this project on HA specific stuff. I'm also CentralCommand in there for reference. I'm pretty familiar with the InfluxDB tests since I wrote or updated nearly all of them so I can answer questions you have on the specifics of those. Might be easier for back and forth while developing. Also pytest does have good documentation, general info on it and its features can be found [here](https://docs.pytest.org/en/6.2.x/).

drlibre committed:
Improved feature for queries_raw option

- "query_raw" config option available in influx V2 mode
- possible mixed configurations "queries" + "queries_raw" or "queries_flux" + queries_raw"

drlibre committed:
Update in accordance with sensor.py modifications

- change LANGUAGE_QUERY_RAW to LANGUAGE_RAW
- add error messages

drlibre committed:
Update in accordance with new sensor feature

- Add new sensor configurations relative to the new feature 'query_raw'.
- Add a test to check field constitency between configuration and query.
- Add _get_sensor(queries) to check configuration with more than one sensor in 'test_minimal_config', 'test_full_config' and 'test_state_matches_query_result'

drlibre committed:
test_sensor.py: update BASE_V1_QUERY_MIXED

Update BASE_V1_QUERY_MIXED to its normal definition

mdegat01 commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -70,34 +75,95 @@
+        return True
+
+    if CONF_QUERIES_RAW in conf:
+        conf[CONF_QUERIES_RAW][:] = [
+            query for query in conf[CONF_QUERIES_RAW] if __is_valid(query)
+        ]
```
You need to call `_merge_connection_config_into_query` on each item in the `queries_raw` array just like it does [here for queries_flux](https://github.com/home-assistant/core/blob/dev/homeassistant/components/influxdb/sensor.py#L89) and [here for queries](https://github.com/home-assistant/core/blob/dev/homeassistant/components/influxdb/sensor.py#L101).

Otherwise it will only work if you provide a value for `database` in each item in `queries_raw`. A config like this:
```yaml
api_version: 1
database: my_default_database
queries_raw:
  - name: Test query 
    query: Test query
```
Won't work right now. `Test query` should inherit the value provided for `database` at the top level since it doesn't override it but since you aren't merging those details in then it won't work.

mdegat01 replied:
Also your code seems to only return invalid if every item in `queries_raw` is invalid. But if there are one or two valid ones and a bunch of invalid ones then it will proceed here. It should not validate a partially valid config, its either all valid or all not.

drlibre replied:
> You need to call `_merge_connection_config_into_query` on each item in the `queries_raw` array just like it does [here for queries_flux](https://github.com/home-assistant/core/blob/dev/homeassistant/components/influxdb/sensor.py#L89) and [here for queries](https://github.com/home-assistant/core/blob/dev/homeassistant/components/influxdb/sensor.py#L101).

Yes, it normally does in `validate_query_format_for_version` as for `queries` or `queries_flux`, here for [queries_raw API V1](https://github.com/home-assistant/core/blob/dev/homeassistant/components/influxdb/sensor.py#L158) and here for [queries_raw API V2](https://github.com/home-assistant/core/blob/dev/homeassistant/components/influxdb/sensor.py#L137) 
          
I had in mind to remove `queries_raw` sensors with invalid configuration due to the configured API mode (for instance `database` in a `queries_raw` sensor whereas influxDB is configured in API V2 mode.
 the need to check there isn't any problem in config for a `queries_raw` sensor, like a `bucket` field whereas influxDB is configured in API V1 mode or `database` whereas influxDB is configured in API V2 mode. I found that more usefull than raising a `vol.Invalid` as this doesn't prevent other sensors with correct config to be setup.

By the way, as `field` and `bucket` options are removed and to be consistent with the general behavior in case of invalid configuration, this part of code will be redifined.

mdegat01 commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -123,6 +189,14 @@ def validate_query_format_for_version(conf: dict) -> dict:
+            vol.Optional(CONF_DB_NAME): cv.string,
+            vol.Optional(CONF_BUCKET): cv.string,
+            vol.Required(CONF_QUERY): cv.template,
+            vol.Optional(CONF_FIELD, default=INFLUX_CONF_VALUE_V2): cv.string,
```
```suggestion
            vol.Optional(CONF_FIELD, default=DEFAULT_FIELD): cv.string,
```
I would not default to the V2 field here. I would default this to the V1 field (which is `"value"` looks like). For V2 just completely ignore this config option and always use `_value`. We'll need to make sure the doc notes that this field is only for V1.

drlibre replied:
OK. As the `CONF_FIELD` won't be valid for V2, the default value for V1 will be set up in the `validate_raw_version_config` function.

mdegat01 commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -163,13 +239,28 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+                else:
+                    _LOGGER.error(NO_DATABASE_ERROR, query[CONF_DB_NAME])
+
+            if query[CONF_API_VERSION] == API_VERSION_2:
+                if query[CONF_BUCKET] in influx.data_repositories:
+                    entities.append(InfluxSensor(hass, influx, query))
+                else:
+                    _LOGGER.error(NO_BUCKET_ERROR, query[CONF_BUCKET])
+
```
```suggestion
```
Remove this for the same reasons as discussed in other comments. We shouldn't prompt for a bucket with V2, it just goes in the raw flux query.

drlibre replied:
Ok, will be removed

mdegat01 commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -341,3 +441,68 @@ def update(self):
+            else:
+                if len(tables) > 1 or len(tables[0].records) > 1:
+                    _LOGGER.warning(QUERY_MULTIPLE_RESULTS_MESSAGE, self.full_query)
+                self.value = tables[0].records[0].values[self.field]
```
```suggestion
                self.value = tables[0].records[0].values[INFLUX_CONF_VALUE_V2]
```
Always use `_value` for V2.

drlibre replied:
OK.

mdegat01 commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -341,3 +441,68 @@ def update(self):
+            if not tables:
+                _LOGGER.warning(QUERY_NO_RESULTS_MESSAGE, self.full_query)
+                self.value = None
+            elif self.field not in tables[0].records[0].values:
+                _LOGGER.warning(QUERY_FIELD_CONSISTENCY, self.field)
+                self.value = None
```
Not really needed for V2 since it should always be `_value`. If you want to put it in just in case then we'll have to change the message to something like "Response was missing an `_value` field, please check your query".

drlibre replied:
OK.

mdegat01 reviewed (changes requested):
This is looking good. I know there's a bunch of changes but tried to put in suggestions where I could easily. And thanks for adding the tests.

drlibre commented on tests/components/influxdb/test_sensor.py:
```diff
@@ -272,16 +304,169 @@ async def test_minimal_config(hass, mock_client, config_ext, queries, set_query_
+            },
+            _set_query_mock_v1,
+        ),
+        (
+            API_VERSION_2,
+            {
+                "api_version": "2",
+                "ssl": "true",
+                "host": "host",
+                "port": "9000",
+                "path": "path",
+                "token": "token",
+                "organization": "org",
+                "bucket": "bucket",
+            },
+            {
+                "queries_flux": [
+                    {
+                        "name": "test",
+                        "unit_of_measurement": "unit",
+                        "range_start": "start",
+                        "range_stop": "end",
+                        "group_function": "fn",
+                        "bucket": "bucket2",
+                        "imports": "import",
+                        "query": "query",
+                    }
+                ],
+                "queries_raw": [
+                    {
+                        "name": "test_r",
+                        "unit_of_measurement": "unit",
+                        "bucket": "bucket2",
+                        "query": "query",
+                        "field": "field",
+                    }
+                ],
+            },
+            _set_query_mock_v2,
+        ),
```
OK, I will add a new `test_queries_raw_combination` for the mixed config in the next version and leave the 'single' `queries_raw` config for API v1 and API V2 in the `test_full_config`

drlibre commented on tests/components/influxdb/test_sensor.py:
```diff
@@ -291,6 +476,45 @@ async def test_config_failure(hass, config_ext):
+    ],
+    indirect=["mock_client"],
+)
+async def test_state_for_unconsistent_field(
```
Right, I'll change in the next version...

drlibre commented on tests/components/influxdb/test_sensor.py:
```diff
@@ -308,6 +532,34 @@ async def test_config_failure(hass, config_ext):
             _set_query_mock_v2,
             _make_v2_resultset,
         ),
+        (
+            DEFAULT_API_VERSION,
+            BASE_V1_CONFIG,
+            BASE_V1_QUERY_RAW,
+            _set_query_mock_v1,
+            _make_v1_resultset,
+        ),
```
Thanks for explanations. 
If not necessary, I'll remove the mixed configs and so, it' simpler!

drlibre commented on tests/components/influxdb/test_sensor.py:
```diff
@@ -470,6 +796,30 @@ async def test_error_querying_influx(
             _make_v2_resultset,
             "query",
         ),
+        (
+            DEFAULT_API_VERSION,
+            BASE_V1_CONFIG,
+            {
+                "queries_raw": [
+                    {
+                        "name": "test",
+                        "query": "{{ illegal.template }}",
+                        "field": "value",
+                    }
+                ]
+            },
+            _set_query_mock_v1,
+            _make_v1_resultset,
+            "query",
+        ),
```
If not necessary, no reason to keep it... Will be removed in the next version.

drlibre commented on tests/components/influxdb/test_sensor.py:
```diff
@@ -537,6 +887,46 @@ async def test_error_rendering_template(
+        (
+            DEFAULT_API_VERSION,
+            BASE_V1_CONFIG,
+            BASE_V1_QUERY_RAW,
+            _set_query_mock_v1,
+            OSError("fail"),
+            _make_v1_resultset,
+        ),
+        (
+            DEFAULT_API_VERSION,
+            BASE_V1_CONFIG,
+            BASE_V1_QUERY_RAW,
+            _set_query_mock_v1,
+            InfluxDBClientError("fail"),
+            _make_v1_resultset,
+        ),
+        (
+            DEFAULT_API_VERSION,
+            BASE_V1_CONFIG,
+            BASE_V1_QUERY_RAW,
+            _set_query_mock_v1,
+            InfluxDBServerError("fail"),
+            _make_v1_resultset,
+        ),
+        (
+            API_VERSION_2,
+            BASE_V2_CONFIG,
+            BASE_V2_QUERY_RAW,
+            _set_query_mock_v2,
+            OSError("fail"),
+            _make_v2_resultset,
+        ),
+        (
+            API_VERSION_2,
+            BASE_V2_CONFIG,
+            BASE_V2_QUERY_RAW,
+            _set_query_mock_v2,
+            ApiException(),
+            _make_v2_resultset,
+        ),
```
Thanks for advice. If not necessary, no reason to keep it... Will be removed in the next version.

drlibre commented on tests/components/influxdb/test_sensor.py:
```diff
@@ -590,6 +980,23 @@ async def test_connection_error_at_startup(
             BASE_V2_QUERY,
             _set_query_mock_v2,
         ),
+        (
+            DEFAULT_API_VERSION,
+            {"database": "bad_db"},
+            BASE_V1_QUERY_RAW,
+            _set_query_mock_v1,
+        ),
+        (
+            API_VERSION_2,
+            {
+                "api_version": API_VERSION_2,
+                "organization": "org",
+                "token": "token",
+                "bucket": "bad_bucket",
+            },
+            BASE_V2_QUERY_RAW,
+            _set_query_mock_v2,
+        ),
```
Ok, removed in the next version.

drlibre commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -123,6 +189,14 @@ def validate_query_format_for_version(conf: dict) -> dict:
+    LANGUAGE_RAW: _QUERY_SENSOR_SCHEMA.extend(
+        {
+            vol.Optional(CONF_DB_NAME): cv.string,
+            vol.Optional(CONF_BUCKET): cv.string,
```
Ok

drlibre commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -70,34 +75,95 @@
+            )
+            return False
+
+        if conf[CONF_API_VERSION] == DEFAULT_API_VERSION and CONF_BUCKET in setting:
+            _LOGGER.error(
+                QUERIES_RAW_BASE_ERROR,
+                setting.get(CONF_NAME),
+                CONF_DB_NAME,
+                CONF_API_VERSION,
+                DEFAULT_API_VERSION,
+            )
+            return False
+
```
OK.

drlibre commented on tests/components/influxdb/test_sensor.py:
```diff
@@ -204,13 +232,17 @@ async def _setup(hass, config_ext, queries, expected_sensors):
     [
         (DEFAULT_API_VERSION, BASE_V1_CONFIG, BASE_V1_QUERY, _set_query_mock_v1),
         (API_VERSION_2, BASE_V2_CONFIG, BASE_V2_QUERY, _set_query_mock_v2),
+        (DEFAULT_API_VERSION, BASE_V1_CONFIG, BASE_V1_QUERY_RAW, _set_query_mock_v1),
+        (API_VERSION_2, BASE_V2_CONFIG, BASE_V2_QUERY_RAW, _set_query_mock_v2),
+        (DEFAULT_API_VERSION, BASE_V1_CONFIG, BASE_V1_QUERY_MIXED, _set_query_mock_v1),
+        (API_VERSION_2, BASE_V2_CONFIG, BASE_V2_QUERY_MIXED, _set_query_mock_v2),
```
The `test_queries_raw_combination` probably doesn't need to test both minimal and maximal conf in mixed format. I suggest to test only minimal conf.

drlibre committed:
Update taking into account comments

drlibre committed:
Update test_sensor.py

Restore initial state of the test

drlibre committed:
Update test_sensor.py

mdegat01 commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -163,13 +217,25 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+                    _LOGGER.error(NO_DATABASE_ERROR, query[CONF_DB_NAME])
+
+            if query[CONF_API_VERSION] == API_VERSION_2:
+                entities.append(InfluxSensor(hass, influx, query))
```
```suggestion
            if query[CONF_BUCKET] in influx.data_repositories:
                entities.append(InfluxSensor(hass, influx, query))
            else:
                _LOGGER.error(NO_BUCKET_ERROR, query[CONF_BUCKET])
```
Same check as in the queries flux part above

drlibre replied:
Your right, many thanks for your review!

mdegat01 reviewed (changes requested):
Very small tweak, need to do the same bucket exists check for v2 in queries raw that we do normally. That and it looks like it needs a rebase to catch up to current dev. But otherwise ready to go!

Really sorry for taking so long on this @drlibre . Got busy with other stuff and then kind of lost track of it in my backlog.

drlibre committed:
Update homeassistant/components/influxdb/sensor.py

Add bucket exists check for v2 in queries raw

Co-authored-by: Mike Degatano <michael.degatano@gmail.com>

mdegat01 commented:
@drlibre almost ready you just need to rebase and resolve the conflict with the test_sensor file

drlibre committed:
Merge branch 'dev' into drlibre-2021-01

drlibre committed:
Merge branch 'dev' into drlibre-2021-01

mdegat01 commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -173,13 +227,28 @@ def setup_platform(
+                    _LOGGER.error(NO_DATABASE_ERROR, query[CONF_DB_NAME])
+
+            if query[CONF_API_VERSION] == API_VERSION_2:
+            if query[CONF_BUCKET] in influx.data_repositories:
+                entities.append(InfluxSensor(hass, influx, query))
+            else:
+                _LOGGER.error(NO_BUCKET_ERROR, query[CONF_BUCKET])
```
Indentation issue, just indent all this one level more.

drlibre replied:
Yes, sorry not to havec done it before... I commit a new version with the correction.

mdegat01 reviewed (changes requested):
Once the indentation issue is fixed this is good. I just tested it locally and confirmed it works as well.

drlibre committed:
Indentation correction line 247-250

drlibre committed:
query_raw test parameter corrections

drlibre committed:
Remove useless field test for API2 Query_raw

drlibre committed:
Test code cover: add 3 cases and modify 1 test

drlibre committed:
Merge branch 'dev' into drlibre-2021-01

drlibre committed:
Merge branch 'dev' into drlibre-2021-01

edenhaus reviewed (changes requested):
Can you please rebase and fix the merge conflicts

Thanks :)

drlibre commented:
> Can you please rebase and fix the merge conflicts
> 
> Thanks :)

Thanks for your comment. 
I tried to set up again my dev environnement (following instructions of https://developers.home-assistant.io/docs/development_environment/) but I always face an error that abort the process...

drlibre committed:
Merge branch 'dev' into drlibre-2021-01

drlibre commented:
Resolved conflict (took into account #88165 & #87996 modifications)

drlibre committed:
Merge branch 'dev' into drlibre-2021-01

drlibre committed:
Merge branch 'dev' into drlibre-2021-01

drlibre committed:
Merge branch 'dev' into drlibre-2021-01

drlibre reviewed (commented):
ok

drlibre reviewed (commented):
ok

drlibre committed:
Correction for Black and Ruff tests

drlibre committed:
Merge branch 'dev' into drlibre-2021-01

drlibre committed:
Merge branch 'dev' into drlibre-2021-01

drlibre committed:
Merge branch 'dev' into drlibre-2021-01

drlibre committed:
Merge branch 'home-assistant:dev' into drlibre-2021-01

edenhaus commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -174,13 +225,28 @@ def setup_platform(
                 _LOGGER.error(NO_BUCKET_ERROR, query[CONF_BUCKET])
-    else:
+
+    if CONF_QUERIES in config:
```
```suggestion
    elif CONF_QUERIES in config:
```

`CONF_QUERIES_FLUX` and `CONF_QUERIES` are exclusiv

edenhaus commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -174,13 +225,28 @@ def setup_platform(
+                else:
+                    _LOGGER.error(NO_DATABASE_ERROR, query[CONF_DB_NAME])
+
+            if query[CONF_API_VERSION] == API_VERSION_2:
```
```suggestion
            else:
```

edenhaus commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -146,8 +195,10 @@ def validate_query_format_for_version(conf: dict) -> dict:
         {
             vol.Exclusive(CONF_QUERIES, "queries"): [_QUERY_SCHEMA[LANGUAGE_INFLUXQL]],
             vol.Exclusive(CONF_QUERIES_FLUX, "queries"): [_QUERY_SCHEMA[LANGUAGE_FLUX]],
+            vol.Optional(CONF_QUERIES_RAW, "queries"): [_QUERY_SCHEMA[LANGUAGE_RAW]],
```
```suggestion
            vol.Optional(CONF_QUERIES_RAW): [_QUERY_SCHEMA[LANGUAGE_RAW]],
```

edenhaus commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -215,7 +281,16 @@ def __init__(self, hass, influx, query):
+        elif query[CONF_LANGUAGE] == LANGUAGE_RAW:
+            query_clause = query.get(CONF_QUERY)
+            query_clause.hass = hass
+            self.data = InfluxRawSensorData(
+                influx,
+                query.get(CONF_API_VERSION),
+                query.get(CONF_DB_NAME),
+                query_clause,
+                query.get(CONF_FIELD),
+            )
```
```suggestion
            if query[CONF_API_VERSION] == DEFAULT_API_VERSION:
                self.data = InfluxV1RawSensorData(
                    influx, query_clause, query.get(CONF_DB_NAME), query.get(CONF_FIELD)
                )
            else:
                self.data = InfluxV2RawSensorData(
                    influx,
                    query_clause,
                )
```

Instead of checking the API version on each update call, we do it once in the init function.

edenhaus commented on homeassistant/components/influxdb/sensor.py:
```diff
@@ -359,3 +434,65 @@ def update(self):
+            _LOGGER.error(RENDERING_QUERY_ERROR_MESSAGE, ex)
+            return
+
+        self.full_query = f"{rendered_query}"
+
+        _LOGGER.debug(RUNNING_QUERY_MESSAGE, self.full_query)
+
+        if self.api_version == DEFAULT_API_VERSION:
+            try:
+                points = self.influx.query(self.full_query, self.db_name)
+            except (ConnectionError, ValueError) as exc:
+                _LOGGER.error(exc)
+                self.value = None
+                return
+
+            if not points:
+                _LOGGER.warning(QUERY_NO_RESULTS_MESSAGE, self.full_query)
+                self.value = None
+            elif self.field not in points[0]:
+                _LOGGER.warning(QUERY_FIELD_CONSISTENCY, self.field)
+                self.value = None
+            else:
+                if len(points) > 1:
+                    _LOGGER.warning(QUERY_MULTIPLE_RESULTS_MESSAGE, self.full_query)
+                self.value = points[0].get(self.field)
+        else:
+            try:
+                tables = self.influx.query(self.full_query)
+            except (ConnectionError, ValueError) as exc:
+                _LOGGER.error(exc)
+                self.value = None
+                return
+
+            if not tables:
+                _LOGGER.warning(QUERY_NO_RESULTS_MESSAGE, self.full_query)
+                self.value = None
+            else:
+                if len(tables) > 1 or len(tables[0].records) > 1:
+                    _LOGGER.warning(QUERY_MULTIPLE_RESULTS_MESSAGE, self.full_query)
+                self.value = tables[0].records[0].values[INFLUX_CONF_VALUE_V2]
```
```suggestion
class AbstractInfluxSensorData(ABC):
    """Abstract Influx sensor data."""

    def __init__(self, influx: InfluxClient, query) -> None:
        """Initialize the data object."""
        self._influx = influx
        self._query = query
        self.value = None

    @Throttle(MIN_TIME_BETWEEN_UPDATES)
    def update(self) -> None:
        """Get the latest data by querying influx."""
        _LOGGER.debug(RENDERING_QUERY_MESSAGE, self.query)
        try:
            rendered_query = self.query.render(parse_result=False)
        except TemplateError as ex:
            _LOGGER.error(RENDERING_QUERY_ERROR_MESSAGE, ex)
            return

        self.value = self._update_specific(rendered_query)

    @abstractmethod
    def _update_specific(self, rendered_query) -> Any:
        """Handle api specific api code."""


class InfluxV1RawSensorData(AbstractInfluxSensorData):
    """Class for handling the data retrieval with v1 API and queries_raw format."""

    def __init__(self, influx: InfluxClient, query, db_name, field):
        """Initialize the data object."""
        super().__init__(influx, query)
        self._db_name = db_name
        self._field = field

    def _update_specific(self, rendered_query) -> Any:
        """Handle api specific api code."""
        full_query = f"{rendered_query}"

        _LOGGER.debug(RUNNING_QUERY_MESSAGE, full_query)

        try:
            points = self._influx.query(full_query, self._db_name)
        except (ConnectionError, ValueError) as exc:
            _LOGGER.error(exc)
            return None

        if not points:
            _LOGGER.warning(QUERY_NO_RESULTS_MESSAGE, full_query)
            return None
        elif self._field not in points[0]:
            _LOGGER.warning(QUERY_FIELD_CONSISTENCY, self._field)
            return None
        else:
            if len(points) > 1:
                _LOGGER.warning(QUERY_MULTIPLE_RESULTS_MESSAGE, full_query)
            return points[0].get(self._field)


class InfluxV2RawSensorData(AbstractInfluxSensorData):
    """Class for handling the data retrieval with v2 API and queries_raw format."""

    def _update_specific(self, rendered_query) -> Any:
        """Handle api specific api code."""
        full_query = f"{rendered_query}"

        _LOGGER.debug(RUNNING_QUERY_MESSAGE, full_query)

        try:
            tables = self._influx.query(full_query)
        except (ConnectionError, ValueError) as exc:
            _LOGGER.error(exc)
            return None

        if not tables:
            _LOGGER.warning(QUERY_NO_RESULTS_MESSAGE, full_query)
            return None

        if len(tables) > 1 or len(tables[0].records) > 1:
            _LOGGER.warning(QUERY_MULTIPLE_RESULTS_MESSAGE, full_query)
        return tables[0].records[0].values[INFLUX_CONF_VALUE_V2]
```

Split the API specific code into two different classes
